code/modules/dwarven/dwarven_structures.dm

```diff
@@ -0,0 +1,428 @@
+	new wanted_pickaxe(drop_location())
+	user?.mind.adjust_experience(/datum/skill/operating, 4)
+
+///Handles the choosing and creation of upgrade kits
+/obj/structure/destructible/dwarven/workshop/proc/handle_upgrades(mob/user)
+	var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
+
+	if(!materials.has_enough_of_material(10000,/datum/material/dwarven))
+		return
+	materials.use_materials(list(/datum/material/dwarven = 10000))
+	new /obj/item/dwarven/upgrade_kit(drop_location())
+	user?.mind.adjust_experience(/datum/skill/operating, 4)
+
+
+/obj/structure/destructible/dwarven/anvil
+	name = "Dwarven anvil"
+	icon_state = "anvil"
+	var/state = 0
+	var/list/obj/item/crafting_list = list(
+		/obj/item/clothing/suit/armor/vest/dwarven_platemail = 20000,
+		/obj/item/clothing/suit/armor/vest/dwarven_chainmail  = 10000,
+		/obj/item/clothing/head/helmet/dwarven_helmet  = 10000,
+		/obj/item/twohanded/war_hammer = 20000,
+		/obj/item/hatchet/dwarven/axe = 10000,
+		/obj/item/hatchet/dwarven/javelin = 10000
+		)
+	custom_materials = list(/datum/material/dwarven = 20000)
+	var/static/list/allowed_types = list(
+		/datum/material/iron,
+		/datum/material/glass,
+		/datum/material/gold,
+		/datum/material/silver,
+		/datum/material/diamond,
+		/datum/material/uranium,
+		/datum/material/plasma,
+		/datum/material/bluespace,
+		/datum/material/bananium,
+		/datum/material/titanium,
+		/datum/material/runite,
+		/datum/material/plastic,
+		/datum/material/adamantine,
+		/datum/material/mythril,
+		/datum/material/wood,
+		/datum/material/dwarven,
+		)
+	//temporary solution til alloys become its own datum.
+	var/static/list/banned_alloys = list(
+		/obj/item/stack/sheet/mineral/plastitanium,
+		/obj/item/stack/sheet/plasteel,
+		/obj/item/stack/sheet/plasmaglass,
+		/obj/item/stack/sheet/rglass,
+		/obj/item/stack/sheet/plasmarglass,
+		/obj/item/stack/sheet/titaniumglass,
+		/obj/item/stack/sheet/plastitaniumglass
+		)
+	var/datum/material/current_mat
+
+/obj/structure/destructible/dwarven/anvil/Initialize()
+	. = ..()
+	//temporary solution til alloys become its own datum.
+	var/list/allowed_materials = list()
+	for(var/i in typecacheof(/obj/item/stack))
+		if(!(i in banned_alloys))
+			allowed_materials += i
+
+	AddComponent(/datum/component/material_container,allowed_types, 200000, TRUE, allowed_materials ,null, _after_insert=CALLBACK(src, .proc/AfterMaterialInsert))
+
+/obj/structure/destructible/dwarven/anvil/AltClick(mob/user)
+	if(!istype(user) || !user.canUseTopic(src, BE_CLOSE))
+		return
+	if(state == 5)
+		state = 0
+	else
+		state++
+
+/obj/structure/destructible/dwarven/anvil/proc/AfterMaterialInsert(_item_inserted, id_inserted, amount_inserted)
+	var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
+	for(var/datum/material/M in materials.materials)
+		if(M != id_inserted)
+			materials.retrieve_sheets(materials.materials[M],M,drop_location())
```

Comment:
this is so fucked in playtesting, anyone can override your mats mid construction